js/src/index.ts

```diff
@@ -144,13 +233,87 @@ export class Schematic {
 
       // After setting context and getting a response, return the value
       const contextVals = this.checks[contextStr] ?? {};
-      return contextVals[key]?.value ?? fallback;
+      const flagCheck = contextVals[key];
+      const result = flagCheck?.value ?? fallback;
+
+      this.debug(
+        `checkFlag WebSocket result: ${key}`,
+        typeof flagCheck !== "undefined"
+          ? flagCheck
+          : { value: fallback, fallbackUsed: true },
+      );
+
+      // If we have flag check results, submit an event
+      if (typeof flagCheck !== "undefined") {
+        this.submitFlagCheckEvent(key, flagCheck, context);
+      }
+
+      return result;
     } catch (error) {
       console.error("Unexpected error in checkFlag:", error);
+
+      // Create a minimal result for the error case and submit event
+      const errorResult: CheckFlagReturn = {
+        flag: key,
+        value: fallback,
+        reason: "Unexpected error in flag check",
+        error: error instanceof Error ? error.message : String(error),
+      };
+      this.submitFlagCheckEvent(key, errorResult, context);
+
       return fallback;
     }
   }
 
+  /**
+   * Helper function to log debug messages
+   * Only logs if debug mode is enabled
+   */
+  private debug(message: string, ...args: unknown[]): void {
+    if (this.debugEnabled) {
+      console.log(`[Schematic] ${message}`, ...args);
+    }
+  }
+
+  /**
+   * Helper function to check if client is in offline mode
+   * Logs debug message about offline mode if debug is enabled
+   */
+  private isOffline(): boolean {
+    if (this.offlineEnabled) {
+      this.debug("Offline mode enabled, skipping network request");
+      return true;
```

Comment:
i feel like this will print a lot, do we want this printing on every isOffline()?

I feel like once on load should be enough

Reply:
yeah I'll remove this, you're right